tests/unit/coroutines_test.cc

```diff
@@ -1018,3 +1020,121 @@ SEASTAR_TEST_CASE(test_lambda_coroutine_in_continuation) {
     }));
     BOOST_REQUIRE_EQUAL(sin1, sin2);
 }
+
+class test_exception : std::exception { };
+class handled_test_exception : std::exception { };
+
+future<> throw_void() {
+    co_await sleep(1ms);
+    throw test_exception{};
+}
+
+future<> return_void() {
+    co_await sleep(1ms);
+}
+
+future<int> throw_int() {
+    co_await sleep(1ms);
+    throw test_exception{};
+}
+
+future<int> return_int() {
+    co_await sleep(1ms);
+    co_return 128;
+}
+
+class dummy {
+    int& _c;
+
+public:
+    explicit dummy(int& c) : _c(c) { ++_c; }
+    dummy(const dummy& o) : _c(o._c) { ++_c; }
+    dummy(dummy&&) = delete;
+    ~dummy() { --_c; }
+};
+
+using try_future_exception_handler_fn = seastar::noncopyable_function<std::exception_ptr(std::exception_ptr)>;
+
+auto do_run_try_future_test(auto f, try_future_exception_handler_fn ex_handler, int& ctor_dtor_counter, bool& run_past)
```

Comment:
Constraint? Guessing std::invocable<>?

Also give it a better name, like return_failed_future.